### PR TITLE
[ao][sparsity] Basic implementation of activation sparsifier

### PR DESCRIPTION
**Pull request by macandro96**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81814
* #80890
* #80889
* #80888
* #80887
* __->__ #80886

The Activation sparsifier class aims to sparsify/prune activations in a neural
network. The idea is to attach the sparsifier to a layer (or layers) and it
zeroes out the activations based on the mask_fn (or sparsification function)
input by the user.
The mask_fn is applied once all the inputs are aggregated and reduced i.e.
mask = mask_fn(reduce_fn(aggregate_fn(activations)))

Note::
    The sparsification mask is computed on the input **before it goes through the attached layer**.

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

### CODE REVIEW

File: test/ao/sparsity/test_activation_sparsifier.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+        batch_size = x.shape[0]
+        out = out.reshape(batch_size, -1)
+
+        out = F.relu(self.linear1(out))
+        out = self.linear2(out)
+        return out
+
+
+class _ActivationSparsifierUtils:
```
Comment (z-a-f) [inline test/ao/sparsity/test_activation_sparsifier.py:36]: It looks like this class is just a collection of functions to test. Can we remove it or pass the dict of functions directly? Not sure you need a factory here

File: test/ao/sparsity/test_activation_sparsifier.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        def reduce_fn(x):
+            return torch.mean(x, dim=0)
+
+        return {
+            'aggregate_fn': agg_fn,
+            'reduce_fn': reduce_fn
+        }, sparse_config
+
+
+class _ActivationSparsiferTestCase(TestCase):
```
Comment (z-a-f) [inline test/ao/sparsity/test_activation_sparsifier.py:61]: Is there a reason this is a separate test class? It looks like it is being used only in the TestActivationSparsifier -- why not just using that?

File: test/ao/sparsity/test_activation_sparsifier.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+class TestActivationSparsifier(_ActivationSparsiferTestCase):
+    def test_activation_sparsifier(self):
+        """Simulates the workflow of the activation sparsifier, starting from object creation
+        till squash_mask().
+        The idea is to check that everything works as expected while in the workflow.
+        """
+        # simulate the workflow
+
+        # STEP 1: make data and activation sparsifier object
+        defaults, sparse_config = _ActivationSparsifierUtils._get_defaults()
```
Comment (z-a-f) [inline test/ao/sparsity/test_activation_sparsifier.py:91]: There is no point for the container class -- just pass them directly. That would improve the readability of this file

File: torch/ao/sparsity/experimental/activation_sparsifier/activation_sparsifier.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+    network. The idea is to attach the sparsifier to a layer (or layers) and it
+    zeroes out the activations based on the mask_fn (or sparsification function)
+    input by the user.
+    The mask_fn is applied once all the inputs are aggregated and reduced i.e.
+    mask = mask_fn(reduce_fn(aggregate_fn(activations)))
+
+    Note::
+        The sparsification mask is computed on the input **before it goes through the attached layer**.
+
+    Args:
```
Comment (z-a-f) [inline torch/ao/sparsity/experimental/activation_sparsifier/activation_sparsifier.py:20]: Please, follow the sphinx syntax: https://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html

- No blanks between args
- Args should end with a colon

File: torch/ao/sparsity/experimental/activation_sparsifier/activation_sparsifier.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+            default aggregate_fn that is used if not specified while registering the layer.
+            specifies how inputs should be aggreagated over time.
+
+        reduce_fn (Optional, function type)
+            default reduce_fn that is used if not specified while registering the layer.
+            reduce_fn will be applied on the aggregated input
+
+        mask_fn (Optional, function type)
+            default mask_fn that is used to create the sparsification mask (based on the
+            aggregated & reduced input). This is used if not specified while registering the layer
```
Comment (z-a-f) [inline torch/ao/sparsity/experimental/activation_sparsifier/activation_sparsifier.py:35]: `function type` is often referred to as `Callable`

Comment (z-a-f) [inline torch/ao/sparsity/experimental/activation_sparsifier/activation_sparsifier.py:35]: For all callables it is usually a good idea to include information how that callable will be used / called. That would give information to the developers on what arguments should a function expect.

File: torch/ao/sparsity/experimental/activation_sparsifier/activation_sparsifier.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+        feature_dim (Optional, int)
+            default dimension of input features. Again, features along this dim will be chosen
+            for sparsification.
+
+        sparse_config (Dict)
+            Default configuration for the mask_fn. This config will be passed
+            with the mask_fn()
+
+    Expected Usage:
```
Comment (z-a-f) [inline torch/ao/sparsity/experimental/activation_sparsifier/activation_sparsifier.py:51]: nit: Replace with `Example:`

File: torch/ao/sparsity/experimental/activation_sparsifier/activation_sparsifier.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+        >>> model = SomeModel()
+        >>> act_sparsifier = ActivationSparsifier(...)  # init activation sparsifier
+
```
Comment (z-a-f) [inline torch/ao/sparsity/experimental/activation_sparsifier/activation_sparsifier.py:54]: nit: Avoid blank spaces within the code (unless intended)

**Review by z-a-f — CHANGES_REQUESTED** [review]

File: torch/ao/sparsity/experimental/activation_sparsifier/activation_sparsifier.py [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+    Args:
+        model (nn.Module)
+            The model whose layers will be sparsified. The layers that needs to be
+            sparsified should be added separately using the register_layer() function
+
+        aggregate_fn (Optional, function type)
+            default aggregate_fn that is used if not specified while registering the layer.
+            specifies how inputs should be aggreagated over time.
+
+        reduce_fn (Optional, function type)
```
Comment (HDCharles) [inline torch/ao/sparsity/experimental/activation_sparsifier/activation_sparsifier.py:29]: its hard to understand what the aggregate and reduce_fn do, also if there are defaults, shouldn't those be specified?

File: torch/ao/sparsity/experimental/activation_sparsifier/activation_sparsifier.py [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+        >>> # start training process
+            >>> # epoch starts
+                >>> # model.forward(), compute_loss() and model.backwards()
+            >>> # epoch ends
+            >>> act_sparsifier.step()
+        >>> # end training process
+
+        >>> sparsifier.squash_mask()
+    """
+    def __init__(self, model: nn.Module, aggregate_fn=None, reduce_fn=None, mask_fn=None,
```
Comment (HDCharles) [inline torch/ao/sparsity/experimental/activation_sparsifier/activation_sparsifier.py:79]: should this be passed in more along the lines of the normal sparsifier where a sparse defaults object is passed in?

**Review by HDCharles — APPROVED** [review]
will approve but I feel the API should be brought in line with the normal sparsifier unless there's a good reason.

File: test/ao/sparsity/test_activation_sparsifier.py [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+    """
+    def check_constructor(self, activation_sparsifier, model, defaults, sparse_config):
+        """Check if the model, defaults and sparse_config are loaded correctly in the activation
+        sparsifier
+        """
+        sparsifier_defaults = activation_sparsifier.defaults
+        combined_defaults = {**defaults, 'sparse_config': sparse_config}
+
+        # check model
+        assert activation_sparsifier.model == model
```
Comment (jerryzh168) [inline test/ao/sparsity/test_activation_sparsifier.py:72]: feels like we are checking implementation details here, I feel we do not need to do that

File: test/ao/sparsity/test_activation_sparsifier.py (hunk 1 above)
Comment (macandro96) [reply test/ao/sparsity/test_activation_sparsifier.py]: The idea was to organize the configs into a separate class for better readability purposes. Moved it to main Test Class.

File: torch/ao/sparsity/experimental/activation_sparsifier/activation_sparsifier.py (hunk 4 above)
Comment (macandro96) [reply torch/ao/sparsity/experimental/activation_sparsifier/activation_sparsifier.py:20]: done

File: torch/ao/sparsity/experimental/activation_sparsifier/activation_sparsifier.py (hunk 9 above)
Comment (macandro96) [reply torch/ao/sparsity/experimental/activation_sparsifier/activation_sparsifier.py:80]: So, there are two kinds of config - 
1. function and feature level config: includes aggregate_fn, reduce_fn, mask_fn, features, etc. 
2. sparse config: Sparsity based config. These arguments are passed along with the mask_fn()

The normal sparsifier accepts sparse_config only where as in here we ask user to specify function, feature level config (explicit args) and sparse_config (similar to normal sparsifier).
